README.md
ibus
====

IBus means Intelligent Input Bus. It is an input framework for Linux OS.

1) change default hotkey "win + space"  to "ctrl + space"

2) modify preference page of ibus

3) debug wrong longname when switch input method

4) align dialog of switching input method center

5) remove the table orientation configure item from ibus setup

6) add french\spanish\arabic\russian support



ChangeLog
commit fe321d51ee5f7f4e7f3886ef6f55e4820a92675f
Author: fujiwarat <takao.fujiwara1@gmail.com>
Date:   2013-04-17 09:57:33 +0900

    Release 1.5.2
    
    Review URL: https://codereview.appspot.com/7530044

M	configure.ac

commit 261bc3046f1e5deb5879bac23900ff7dbfa5a20d
Author: Peng Huang <shawn.p.huang@gmail.com>
Date:   2013-03-26 06:55:09 -0400

    Draw highlight for focused IME in IME switcher.
    
    BUG=None
    TEST=Manual
    
    Review URL: https://codereview.appspot.com/7689046

M	ui/gtk3/switcher.vala

commit af1f409a3eace2ea9a063957a8797a50a77d131a
Author: Peng Huang <shawn.p.huang@gmail.com>
Date:   2013-03-26 06:54:43 -0400

    Fix build error and warnings in Fedora 18.
    
    BUG=None
    TEST=Manual
    
    Review URL: https://codereview.appspot.com/7531047

M	bus/dbusimpl.c
M	bus/main.c
M	client/gtk2/ibusimcontext.c
M	configure.ac
M	ibus.spec.in

commit 928351ee88778cf97136d1a7dbc940390c23ddfc
Merge: 175df22 4d7f4ad
Author: Peng Huang <shawn.p.huang@gmail.com>
Date:   2013-03-25 07:29:41 -0700

    Merge pull request #6 from maxiaojun/master
    
    Show all g-ir-scanner warnings

commit 175df2261b431eff65349b47bc6bd086051cde41
Author: fujiwarat <takao.fujiwara1@gmail.com>
Date:   2013-03-21 11:13:52 +0900

    Move ibus.bash script from /etc to /usr/share
    
    BUG=RH#922996
    
    Review URL: https://codereview.appspot.com/7863045

M	tools/Makefile.am

commit 2e7ece775f5412ab8e83a11e4d423b32a6114159
Author: Ville Skyttä <ville.skytta@iki.fi>
Date:   2013-03-12 11:17:59 +0900

    bash completion: No need to install it as executable.
    
    BUG=http://code.google.com/p/ibus/issues/detail?id=1601
    
    Review URL: https://codereview.appspot.com/7696043
    Patch from Ville Skyttä <ville.skytta@iki.fi>.

M	tools/Makefile.am

commit 42dd60fe1fcdce11e7ac517786b398eba0cb9e17
Author: Daiki Ueno <ueno@unixuser.org>
Date:   2013-03-12 06:41:43 +0900

    Add textdomain property to IBusEngineDesc
    
    BUG=Issue#1580
    
    Review URL: https://codereview.appspot.com/7692043

M	bindings/pygobject/gi/overrides/IBus.py
M	src/ibusenginedesc.c
M	src/ibusenginedesc.h

commit 69b48942e8ac9d39afeadfd9eb34f83247e8082a
Author: Daiki Ueno <ueno@unixuser.org>
Date:   2013-03-10 14:55:42 +0900

    Simplify component xml generation using sed
    
    BUG=none
    
    Review URL: https://codereview.appspot.com/7445048

M	conf/dconf/Makefile.am
A	conf/dconf/dconf.xml.in
D	conf/dconf/dconf.xml.in.in
M	conf/gconf/Makefile.am
A	conf/gconf/gconf.xml.in
D	conf/gconf/gconf.xml.in.in
M	conf/memconf/Makefile.am
A	conf/memconf/memconf.xml.in
D	conf/memconf/memconf.xml.in.in
M	configure.ac
M	engine/Makefile.am
M	engine/gensimple.py
A	engine/simple.xml.in
D	engine/simple.xml.in.in
M	ui/gtk2/Makefile.am
A	ui/gtk2/gtkpanel.xml.in
D	ui/gtk2/gtkpanel.xml.in.in
M	ui/gtk3/Makefile.am
A	ui/gtk3/gtkpanel.xml.in
D	ui/gtk3/gtkpanel.xml.in.in

commit cff35929a9330081db25a86a8258489a27d4bdd5
Author: fujiwarat <takao.fujiwara1@gmail.com>
Date:   2013-03-08 11:21:25 +0900

    Launch the second ibus engine too to reduce the launching time.
    
    Currently the first ibus engine only is running but the second engine
    is also frequently used. It is good to run the second engine before
    user types Ctrl+space.
    
    BUG=RH#838734
    TEST=src/tests/ibus-bus
    
    Review URL: https://codereview.appspot.com/7433049

M	bus/ibusimpl.c
M	src/ibusbus.c
M	src/ibusbus.h
M	src/tests/ibus-bus.c
M	ui/gtk3/panel.vala

commit 5fa461889939a91fecff89a67630076b8cb987b2
Author: fujiwarat <takao.fujiwara1@gmail.com>
Date:   2013-03-07 13:16:10 +0900

    Change set_use_system_keyboard_layout() timing.
    
    Call set_use_system_keyboard_layout() before update_engines() is called
    in panel.vala because update_engines() calls switch_engine().
    
    BUG=http://code.google.com/p/ibus/issues/detail?id=1558#c43
    
    Review URL: https://codereview.appspot.com/7559043

M	ui/gtk3/panel.vala

commit 4b2f6a477019b50375aba2e5f9c43e618b6905ad
Author: fujiwarat <takao.fujiwara1@gmail.com>
Date:   2013-03-04 10:10:53 +0900

    Implement use-system-keyboard-layout
    
    BUG=http://code.google.com/p/ibus/issues/detail?id=1558
    
    Review URL: https://codereview.appspot.com/7414045

M	data/ibus.schemas.in
M	ui/gtk3/panel.vala

commit e0878144a704e9bfd0f5a60e6894b1de272aaa7b
Author: Daiki Ueno <ueno@unixuser.org>
Date:   2013-02-27 07:12:09 +0900

    Fix make rpm errors due to ibus-dconf and pygobject override
    
    BUG=Issue#1600
    
    Review URL: https://codereview.appspot.com/7404047

M	ibus.spec.in

commit f9ddda1e2d867b96889f80d9d30c0e88d3ec11b9
Author: Daiki Ueno <ueno@unixuser.org>
Date:   2013-02-26 07:42:14 +0900

    daemon: allow "eavesdrop" match rule
    
    Recent dbus-monitor supplies "eavesdrop" match rule which ibus-daemon
    does not accept.  This patch silently ignores the rule so ibus-daemon
    behave the same as before.
    
    BUG=none
    
    Review URL: https://codereview.appspot.com/7225062

M	bus/matchrule.c
M	bus/test-matchrule.c

commit c79789fa1b5fd49a53b16e87813b88122168ff73
Author: fujiwarat <takao.fujiwara1@gmail.com>
Date:   2013-02-21 12:41:08 +0900

    Fix annotation in ibus_bus_create_input_context
    
    BUG=http://code.google.com/p/ibus/issues/detail?id=1593
    
    Review URL: https://codereview.appspot.com/7368044

M	src/ibusbus.h
M	src/ibusinputcontext.h
M	src/ibuspanelservice.h

commit 4d7f4ad5b5d6b29bb50cc9bf5b1442e3766151d9
Author: Ma Xiaojun <damage3025@gmail.com>
Date:   2013-02-19 19:31:54 +0800

    Show all GIR warnings

M	configure.ac

commit 7e07299d17156e52b0b60b2cc50a82e29ef4f550
Author: fujiwarat <takao.fujiwara1@gmail.com>
Date:   2013-02-15 12:28:25 +0900

    Fix no rule to make target test-override-ibus.py
    
    BUG=http://code.google.com/p/ibus/issues/detail?id=1592
    
    Review URL: https://codereview.appspot.com/7332043

M	bindings/pygobject/Makefile.am

commit 62cd0492e3459416e1775aedc327bced53f66828

src/ibus.h
/* -*- mode: C; c-basic-offset: 4; indent-tabs-mode: nil; -*- */
/* vim:set et sts=4: */
/* ibus - The Input Bus
 * Copyright (C) 2008-2010 Peng Huang <shawn.p.huang@gmail.com>
 * Copyright (C) 2008-2010 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef __IBUS_H_
#define __IBUS_H_

#define __IBUS_H_INSIDE__

#include <ibusversion.h>
#include <ibusshare.h>
#include <ibusobject.h>
#include <ibusserializable.h>
#include <ibusattribute.h>
#include <ibusattrlist.h>
#include <ibustext.h>
#include <ibusproxy.h>
#include <ibusservice.h>
#include <ibusfactory.h>
#include <ibusengine.h>
#include <ibusenginesimple.h>
#include <ibuserror.h>
#include <ibusproperty.h>
#include <ibusproplist.h>
#include <ibuslookuptable.h>
#include <ibusbus.h>
#include <ibuskeysyms.h>
#include <ibuskeymap.h>
#include <ibuskeys.h>
#include <ibusenumtypes.h>
#include <ibushotkey.h>
#include <ibusxkbxml.h>
#include <ibusxml.h>
#include <ibusenginedesc.h>
#include <ibusobservedpath.h>
#include <ibuscomponent.h>
#include <ibusconfig.h>
#include <ibusconfigservice.h>
#include <ibuspanelservice.h>
#include <ibusutil.h>

#ifndef IBUS_DISABLE_DEPRECATED
#include <ibuskeysyms-compat.h>
#endif

#undef __IBUS_H_INSIDE__

#endif



src/ibusversion.h
/* vim:set et sts=4: */
/* ibus - The Input Bus
 * Copyright (C) 2008-2010 Peng Huang <shawn.p.huang@gmail.com>
 * Copyright (C) 2008-2010 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */
/**
 * SECTION: ibusversion
 * @short_description: Current version of IBus.
 * @stability: Stable
 *
 * IBusVersion shows the current IBus version.
 */

#ifndef __IBUS_VERSION_H_
#define __IBUS_VERSION_H_
/* compile time version
 */
/**
 * IBUS_MAJOR_VERSION:
 *
 * IBus major version.
 */
#define IBUS_MAJOR_VERSION				(1)

/**
 * IBUS_MINOR_VERSION:
 *
 * IBus minor version.
 */
#define IBUS_MINOR_VERSION				(5)

/**
 * IBUS_MICRO_VERSION:
 *
 * IBus micro version.
 */
#define IBUS_MICRO_VERSION				(2)

/**
 * IBUS_CHECK_VERSION:
 * @major: Major version of ibus.
 * @minor: Minor version of ibus.
 * @micro: Micro version of ibus.
 *
 * Check whether the current IBus version is equal to or greater than
 * given major.minor.micro.
 */
#define	IBUS_CHECK_VERSION(major,minor,micro)	\
    (IBUS_MAJOR_VERSION > (major) || \
     (IBUS_MAJOR_VERSION == (major) && IBUS_MINOR_VERSION > (minor)) || \
     (IBUS_MAJOR_VERSION == (major) && IBUS_MINOR_VERSION == (minor) && \
      IBUS_MICRO_VERSION >= (micro)))

#endif



setup/enginecombobox.py
# vim:set et sts=4 sw=4:
#
# ibus - The Input Bus
#
# Copyright (c) 2007-2010 Peng Huang <shawn.p.huang@gmail.com>
# Copyright (c) 2007-2010 Red Hat, Inc.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330,
# Boston, MA  02111-1307  USA

import locale

from gi.repository import GObject
from gi.repository import Gtk
from gi.repository import IBus
from gi.repository import Pango

from icon import load_icon
from i18n import _, N_

class EngineComboBox(Gtk.ComboBox):
    __gtype_name__ = 'EngineComboBox'
    __gproperties__ = {
        'active-engine' : (
            object,
            'selected engine',
            'selected engine',
            GObject.ParamFlags.READABLE)
    }

    def __init__(self):
        super(EngineComboBox, self).__init__()
        self.connect("notify::active", self.__notify_active_cb)

        self.__model = None
        self.__all_model = None
        self.__config = None
        self.__show_sub_lang = False

        renderer = Gtk.CellRendererPixbuf()
        renderer.set_property("xalign", 0)
        renderer.set_property("xpad", 2)
        self.pack_start(renderer, False)
        self.set_cell_data_func(renderer, self.__icon_cell_data_cb, None)

        renderer = Gtk.CellRendererText()
        renderer.set_property("xalign", 0)
        renderer.set_property("xpad", 2)
        self.pack_start(renderer, True)
        self.set_cell_data_func(renderer, self.__name_cell_data_cb, None)

    def __gconf_get_lang_list_from_locale(self):
        common_list = ['en', 'Other']
        if  self.__config == None:
            return None
        loc = None
        try:
           loc = locale.setlocale (locale.LC_ALL)
        except:
            pass
        if loc == None:
            return common_list
        current_lang = IBus.get_language_name(loc)
        if current_lang == None:
            return common_list
        group_list = self.__config.get_value("general/xkblayoutconfig",
                                             "group_list")
        if group_list == None:
            return [loc] + common_list
        group_list = list(group_list)
        lang_list = None
        for group in group_list:
            group = str(group)
            langs = list(self.__config.get_value("general/xkblayoutconfig",
                                                  group))
            for lang in langs:
                lang = str(lang)
                if current_lang == IBus.get_language_name(lang):
                    lang_list = langs
                    break
            if lang_list != None:
                break
        if lang_list == None:
            return [loc] + common_list
        return lang_list + common_list

    def __has_engine_in_lang_list(self, engine, lang_list):
        retval = False
        for lang in lang_list:
            if IBus.get_language_name(lang) == \
                IBus.get_language_name(engine.props.language):
                retval = True
                break
        return retval

    def __model_append_langs(self, model, langs, visible):
        keys = langs.keys()
        keys.sort(locale.strcoll)
        loc = locale.getlocale()[0]
        # None on C locale
        if loc == None:
            loc = 'en_US'
        current_lang = IBus.get_language_name(loc)
        # move current language to the first place
        if current_lang in keys:
            keys.remove(current_lang)
            keys.insert(0, current_lang)

        #add "Others" to the end of the combo box
        if IBus.get_language_name("Other") in keys:
            keys.remove(IBus.get_language_name("Other"))
            keys += [IBus.get_language_name("Other")]
        for l in keys:
            #don't append lang
            #iter1 = model.append(None) #iter1, 1level
            #model.set(iter1, 0, l) 
            def cmp_engine(a, b):
                if a.get_rank() == b.get_rank():
                    return locale.strcoll(a.get_longname(), b.get_longname())
                return int(b.get_rank() - a.get_rank())
            langs[l].sort(cmp_engine)
            for e in langs[l]:
                #iter2 = model.append(iter1) #iter2: 2 level
                #model.set(iter2, 0, e)
                iter1 = model.append(None) #iter2: 2 level
                model.set(iter1, 0, e)

    def set_engines(self, engines):
        #model
        self.__model = Gtk.TreeStore(object)

        iter1 = self.__model.append(None)
        self.__model.set(iter1, 0, 0)
        lang_list = self.__gconf_get_lang_list_from_locale()
        lang = {}
        sub_lang = {}
        for e in engines:
            l = IBus.get_language_name(e.props.language)
            if lang_list == None or \
                self.__has_engine_in_lang_list(e, lang_list):
                if l not in lang:
                    lang[l] = []
                lang[l].append(e)
            else:
                if l not in sub_lang:
                    sub_lang[l] = []
                sub_lang[l].append(e)

        self.__model_append_langs(self.__model, lang, True)
        iter1 = self.__model.append(None)
        #self.__model.set(iter1, 0, -1)

        #all model
        self.__all_model = Gtk.TreeStore(object)
        #it will add a column
        #iter1 = self.__all_model.append(None)
        #self.__all_model.set(iter1, 0, 0)
        self.__model_append_langs(self.__all_model, lang, False)
        #remove it ,it will add a column
        #iter1 = self.__all_model.append(None)
        #self.__all_model.set(iter1, 0, -1)
        #ok, it will add other region input method
        self.__model_append_langs(self.__all_model, sub_lang, False)

        self.__toggle_sub_lang()

    def __toggle_sub_lang(self):
        self.set_model(None)

        #echo add start
        #select which model to show, all_model or model
        self.set_model(self.__all_model)
        self.set_active(0)
        return 
        #echo add end

        if self.__show_sub_lang:
            self.set_model(self.__all_model)
        else:
            self.set_model(self.__model)
        self.set_active(0)

    def __icon_cell_data_cb(self, celllayout, renderer, model, iter, data):
        model = self.get_model()
        engine = model.get_value(iter, 0)
        if engine == None :
            return

        #1 level
        if isinstance(engine, str) or isinstance (engine, unicode):
            return
            renderer.set_property("visible", False)
            renderer.set_property("sensitive", False)
        #1 level and top level
        elif isinstance(engine, int):
            if engine == 0:
                renderer.set_property("visible", False)
                renderer.set_property("sensitive", False)
                renderer.set_property("pixbuf", None)
            elif engine < 0:
                return 
                if not self.__show_sub_lang:
                    pixbuf = load_icon("go-bottom", Gtk.IconSize.LARGE_TOOLBAR)
                else:
                    pixbuf = load_icon("go-up", Gtk.IconSize.LARGE_TOOLBAR)
                if pixbuf == None:
                    pixbuf = load_icon(Gtk.STOCK_MISSING_IMAGE,
                                       Gtk.IconSize.LARGE_TOOLBAR)
                if pixbuf == None:
                    renderer.set_property("visible", False)
                    renderer.set_property("sensitive", False)
                    return
                renderer.set_property("visible", True)
                renderer.set_property("sensitive", True)
                renderer.set_property("pixbuf", pixbuf)
        #2 level
        else:
            renderer.set_property("visible", True)
            renderer.set_property("sensitive", True)
            pixbuf = load_icon(engine.get_icon(), Gtk.IconSize.LARGE_TOOLBAR)
            renderer.set_property("pixbuf", pixbuf)

    def __name_cell_data_cb(self, celllayout, renderer, model, iter, data):
        model = self.get_model()
        engine = model.get_value(iter, 0)
        if engine == None :
            return

        #1 level 
        if isinstance (engine, str) or isinstance (engine, unicode):
            renderer.set_property("sensitive", False)
            renderer.set_property("text", engine)
            renderer.set_property("weight", Pango.Weight.NORMAL)
        #1 level and top level
        elif isinstance(engine, int):
            renderer.set_property("sensitive", True)
            if engine == 0:
                renderer.set_property("text", _("Select an input method"))
                renderer.set_property("weight", Pango.Weight.NORMAL)
            elif engine < 0:
                #here
                if not self.__show_sub_lang:
                    renderer.set_property("text", _("Show all input methods"))
                else:
                    renderer.set_property("text", _("Show only input methods for your region"))
                renderer.set_property("weight", Pango.Weight.BOLD)
        # 2 level
        else:
            renderer.set_property("sensitive", True)
            renderer.set_property("text", engine.get_longname())
            renderer.set_property("weight",
                    Pango.Weight.BOLD if engine.get_rank() > 0 else Pango.Weight.NORMAL)

    def __notify_active_cb(self, combobox, property):
        self.notify("active-engine")

    #select a item
    def do_get_property(self, property):
        if property.name == "active-engine":
            i = self.get_active()
            #remove it , otherwise the first item won't be add
            #if i == 0 or i == -1:
            #    return None
            iter = self.get_active_iter()
            model = self.get_model()
            if model[iter][0] == -1:
                self.__show_sub_lang = not self.__show_sub_lang
                self.__toggle_sub_lang()
                return None
            return model[iter][0]
        else:
            raise AttributeError, 'unknown property %s' % property.name

    def set_config(self, config):
        self.__config = config

    def get_active_engine(self):
        return self.get_property("active-engine")

if __name__ == "__main__":
    combo = EngineComboBox()
    combo.set_engines([IBus.EngineDesc(language="zh")])
    w = Gtk.Window()
    w.add(combo)
    w.show_all()
    Gtk.main()


src/ibustypes.h
/* -*- mode: C; c-basic-offset: 4; indent-tabs-mode: nil; -*- */
/* vim:set et sts=4: */
/* ibus - The Input Bus
 * Copyright (C) 2008-2010 Peng Huang <shawn.p.huang@gmail.com>
 * Copyright (C) 2008-2010 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#if !defined (__IBUS_H_INSIDE__) && !defined (IBUS_COMPILATION)
#error "Only <ibus.h> can be included directly"
#endif

/**
 * SECTION: ibustypes
 * @short_description: Generic types for  IBus.
 * @stability: Stable
 *
 * This section consists generic types for IBus, including shift/control key modifiers,
 * and a rectangle structure.
 */
#ifndef __IBUS_TYPES_H_
#define __IBUS_TYPES_H_

/**
 * IBusModifierType:
 * @IBUS_SHIFT_MASK: Shift  is activated.
 * @IBUS_LOCK_MASK: Cap Lock is locked.
 * @IBUS_CONTROL_MASK: Control key is activated.
 * @IBUS_MOD1_MASK: Modifier 1 (Usually Alt_L (0x40),  Alt_R (0x6c),  Meta_L (0xcd)) activated.
 * @IBUS_MOD2_MASK: Modifier 2 (Usually Num_Lock (0x4d)) activated.
 * @IBUS_MOD3_MASK: Modifier 3 activated.
 * @IBUS_MOD4_MASK: Modifier 4 (Usually Super_L (0xce),  Hyper_L (0xcf)) activated.
 * @IBUS_MOD5_MASK: Modifier 5 (ISO_Level3_Shift (0x5c),  Mode_switch (0xcb)) activated.
 * @IBUS_BUTTON1_MASK: Mouse button 1 (left) is activated.
 * @IBUS_BUTTON2_MASK: Mouse button 2 (middle) is activated.
 * @IBUS_BUTTON3_MASK: Mouse button 3 (right) is activated.
 * @IBUS_BUTTON4_MASK: Mouse button 4 (scroll up) is activated.
 * @IBUS_BUTTON5_MASK: Mouse button 5 (scroll down) is activated.
 * @IBUS_HANDLED_MASK: Handled mask indicates the event has been handled by ibus.
 * @IBUS_FORWARD_MASK: Forward mask indicates the event has been forward from ibus.
 * @IBUS_IGNORED_MASK: It is an alias of IBUS_FORWARD_MASK.
 * @IBUS_SUPER_MASK: Super (Usually Win) key is activated.
 * @IBUS_HYPER_MASK: Hyper key is activated.
 * @IBUS_META_MASK: Meta key is activated.
 * @IBUS_RELEASE_MASK: Key is released.
 * @IBUS_MODIFIER_MASK: Modifier mask for the all the masks above.
 *
 * Handles key modifier such as control, shift and alt and release event.
 * Note that nits 15 - 25 are currently unused, while bit 29 is used internally.
 */
typedef enum
{
    IBUS_SHIFT_MASK    = 1 << 0,
    IBUS_LOCK_MASK     = 1 << 1,
    IBUS_CONTROL_MASK  = 1 << 2,
    IBUS_MOD1_MASK     = 1 << 3,
    IBUS_MOD2_MASK     = 1 << 4,
    IBUS_MOD3_MASK     = 1 << 5,
    IBUS_MOD4_MASK     = 1 << 6,
    IBUS_MOD5_MASK     = 1 << 7,
    IBUS_BUTTON1_MASK  = 1 << 8,
    IBUS_BUTTON2_MASK  = 1 << 9,
    IBUS_BUTTON3_MASK  = 1 << 10,
    IBUS_BUTTON4_MASK  = 1 << 11,
    IBUS_BUTTON5_MASK  = 1 << 12,

    /* The next few modifiers are used by XKB, so we skip to the end.
     * Bits 15 - 23 are currently unused. Bit 29 is used internally.
     */

    /* ibus mask */
    IBUS_HANDLED_MASK  = 1 << 24,
    IBUS_FORWARD_MASK  = 1 << 25,
    IBUS_IGNORED_MASK  = IBUS_FORWARD_MASK,

    IBUS_SUPER_MASK    = 1 << 26,
    IBUS_HYPER_MASK    = 1 << 27,
    IBUS_META_MASK     = 1 << 28,

    IBUS_RELEASE_MASK  = 1 << 30,

    IBUS_MODIFIER_MASK = 0x5f001fff
} IBusModifierType;

/**
 * IBusCapabilite:
 * @IBUS_CAP_PREEDIT_TEXT: UI is capable to show pre-edit text.
 * @IBUS_CAP_AUXILIARY_TEXT: UI is capable to show auxiliary text.
 * @IBUS_CAP_LOOKUP_TABLE: UI is capable to show the lookup table.
 * @IBUS_CAP_FOCUS: UI is capable to get focus.
 * @IBUS_CAP_PROPERTY: UI is capable to have property.
 * @IBUS_CAP_SURROUNDING_TEXT: Client can provide surround text,
 *  or IME can handle surround text.
 *
 * Capability flags of UI.
 */
typedef enum {
    IBUS_CAP_PREEDIT_TEXT       = 1 << 0,
    IBUS_CAP_AUXILIARY_TEXT     = 1 << 1,
    IBUS_CAP_LOOKUP_TABLE       = 1 << 2,
    IBUS_CAP_FOCUS              = 1 << 3,
    IBUS_CAP_PROPERTY           = 1 << 4,
    IBUS_CAP_SURROUNDING_TEXT   = 1 << 5,
} IBusCapabilite;

/**
 * IBusPreeditFocusMode:
 * @IBUS_ENGINE_PREEDIT_CLEAR: pre-edit text is cleared.
 * @IBUS_ENGINE_PREEDIT_COMMIT: pre-edit text is committed.
 *
 * Pre-edit commit mode when the focus is lost.
 */
typedef enum {
    IBUS_ENGINE_PREEDIT_CLEAR   = 0,
    IBUS_ENGINE_PREEDIT_COMMIT  = 1,
} IBusPreeditFocusMode;

/**
 * IBusOrientation:
 * @IBUS_ORIENTATION_HORIZONTAL: Horizontal orientation.
 * @IBUS_ORIENTATION_VERTICAL: Vertival orientation.
 * @IBUS_ORIENTATION_SYSTEM: Use ibus global orientation setup.
 *
 * Orientation of UI.
 */
typedef enum {
    IBUS_ORIENTATION_HORIZONTAL = 0,
    IBUS_ORIENTATION_VERTICAL   = 1,
    IBUS_ORIENTATION_SYSTEM     = 2,
} IBusOrientation;

/**
 * IBusBusNameFlag:
 * @IBUS_BUS_NAME_FLAG_ALLOW_REPLACEMENT:
 *    same as DBUS_NAME_FLAG_ALLOW_REPLACEMENT	
 * @IBUS_BUS_NAME_FLAG_REPLACE_EXISTING:
 *    same as DBUS_NAME_FLAG_REPLACE_EXISTING
 * @IBUS_BUS_NAME_FLAG_DO_NOT_QUEUE:
 *    same as DBUS_NAME_FLAG_DO_NOT_QUEUE 
 */
typedef enum {
    IBUS_BUS_NAME_FLAG_ALLOW_REPLACEMENT   = (1 << 0),
    IBUS_BUS_NAME_FLAG_REPLACE_EXISTING    = (1 << 1),
    IBUS_BUS_NAME_FLAG_DO_NOT_QUEUE        = (1 << 2),
} IBusBusNameFlag;

/**
 * IBusBusRequestNameReply:
 * @IBUS_BUS_REQUEST_NAME_REPLY_PRIMARY_OWNER:
 *    same as DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER
 * @IBUS_BUS_REQUEST_NAME_REPLY_IN_QUEUE:
 *    same as DBUS_REQUEST_NAME_REPLY_IN_QUEUE
 * @IBUS_BUS_REQUEST_NAME_REPLY_EXISTS:
 *    same as DBUS_REQUEST_NAME_REPLY_EXISTS
 * @IBUS_BUS_REQUEST_NAME_REPLY_ALREADY_OWNER:
 *    same as DBUS_REQUEST_NAME_REPLY_ALREADY_OWNER
 */
typedef enum {
    IBUS_BUS_REQUEST_NAME_REPLY_PRIMARY_OWNER   = 1,
    IBUS_BUS_REQUEST_NAME_REPLY_IN_QUEUE        = 2,
    IBUS_BUS_REQUEST_NAME_REPLY_EXISTS          = 3,
    IBUS_BUS_REQUEST_NAME_REPLY_ALREADY_OWNER   = 4,
} IBusBusRequestNameReply;

/**
 * IBusBusStartServiceByNameReply:
 * @IBUS_BUS_START_REPLY_SUCCESS:
 *    same as DBUS_START_REPLY_SUCCESS
 * @IBUS_BUS_START_REPLY_ALREADY_RUNNING:
 *    same as DBUS_START_REPLY_ALREADY_RUNNING
 */
typedef enum {
    IBUS_BUS_START_REPLY_SUCCESS = 1,
    IBUS_BUS_START_REPLY_ALREADY_RUNNING = 2,
} IBusBusStartServiceByNameReply;

/**
 * IBusError:
 * @IBUS_ERROR_NO_ENGINE:
 * There is no engine associated with input context.
 * @IBUS_ERROR_NO_CONFIG:
 * There is no config module running.
 * @IBUS_ERROR_FAILED:
 * General failure.
 */
typedef enum {
    IBUS_ERROR_NO_ENGINE,
    IBUS_ERROR_NO_CONFIG,
    IBUS_ERROR_FAILED
} IBusError;

/**
 * IBusPreloadEngineMode:
 * @IBUS_PRELOAD_ENGINE_MODE_USER: user custimized engines
 * @IBUS_PRELOAD_ENGINE_MODE_LANG_RELATIVE: language related engines.
 */
typedef enum {
    IBUS_PRELOAD_ENGINE_MODE_USER          = 0,
    IBUS_PRELOAD_ENGINE_MODE_LANG_RELATIVE = 1,
} IBusPreloadEngineMode;

/**
 * IBusRectangle:
 * @x: x coordinate.
 * @y: y coordinate.
 * @width: width of the rectangle.
 * @height: height of the renctangl.
 *
 * Rectangle definition.
 */
typedef struct _IBusRectangle IBusRectangle;
struct _IBusRectangle {
    gint x;
    gint y;
    gint width;
    gint height;
};

/**
 * IBusFreeFunc:
 * @object: object to be freed.
 *
 * Free function prototype.
 */
typedef void (* IBusFreeFunc) (gpointer object);

#endif

